---
title: "Ampliseq: add phenotype data"
author: "Alexey Larionov"
date: "28 Feb 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary  

The source Ampliseq phenotypes data was provided in several iterations, as the following:  

- Initial phenotype data were provided by Xiaolin in e-mail to Marc of 29Aug2018  
    + The original file name was **Phenotype data for exome seq_shipping manifest_sent_10 05 17.xlsx**  
    + This data is used here as file **phenotypes_dec2018.txt**  
- Then, updated data were provided to me by Marc in e-mail of 07Jan2019  
    + The original file name was **2Phenotype data for exome seq_shipping manifest_sent_10 05 17.xlsx**  
    + This data is used here as file **phenotypes_jan2019.txt**  
    + contains additional data about familial history, X-ray treatment etc  
- Finally, information about Danish samples was sent by Xiaolin to Marc on 11Sep2019  
    + The original file name was **Danish_samples_in_WES.xlsx**


In e-mail of 27Apr2020 Marc clarified that none of clinically actionable genes can be analysed in Danish cohort.  
So, it was decided that these samples should be excluded from this analysis because it aggregates known BC genes.  

and partially prepared during the previous iteration of Ampliseq analysis as the following

- The script read both sets and compared them before use  
- Then the script synced phenotype cases IDs with the samples names in sequencing  

? Then script removed cases failed sequencing from phenotype data and renamed some columns  

The scripts for the previous iteration of Ampliseq analysis of 2019 are avaialbe here ...  

In 2020 it became clear thet we should not use Danish samples to analyse the known Breast cancer predisposition genes.  

The list of such genes was provided by ... on ...  

The current phenotype annotations were compiled in ...  
/Users/alexey/Documents/wecare/Danish_samples/data/updated/ampliseq

Input and Output data: 3,758 vars x 535 samples (265UBC + 270CBC)  

# start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
#options(width = 999)
```

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors = F,
        warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

# Files and folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_wo_danish_2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","s09_add_phenotypes")
#dir.create(data_folder)
scripts_folder <- file.path(base_folder,"scripts","s09_add_phenotypes")
setwd(scripts_folder)

# Libraries
library(dplyr)

```

# Load data  

```{r}

load(file.path(base_folder,"data","s05_phenotypes","add_ampliseq_danish_samples.RData"))

# Update folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_wo_danish_2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","s09_add_phenotypes")
scripts_folder <- file.path(base_folder,"scripts","s09_add_phenotypes")
str
```

```{r}

# Phenotype data
phenotypes_2018_file <- paste(base_folder, "phenotypes_txt/phenotypes_dec2018.txt", sep="/")
phenotypes_2018.df <- read.table(phenotypes_2018_file, header=T, sep="\t", quote="")

phenotypes_2019_file <- paste(base_folder, "phenotypes_txt/phenotypes_jan2019.txt", sep="/")
phenotypes_2019.df <- read.table(phenotypes_2019_file, header=T, sep="\t", quote="")

# Sequencing data
data_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s08_reshape_variants_annotations"
load(paste(data_folder, "s03_update_variants_annotations.RData", sep="/"))
base_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s09_add_phenotypes"

# Clean-up
rm(phenotypes_2018_file, phenotypes_2019_file, data_folder)

```

# Compare phenotypes provided in 2018 and 2019

All fields from 2018 are included in 2019  
Additional fields in 2019: family_history, setno, xrtbreast  

```{r compare_phenotypes}

dim(phenotypes_2018.df)
dim(phenotypes_2019.df)

sum(is.na(phenotypes_2018.df))
sum(is.na(phenotypes_2019.df))

colnames(phenotypes_2018.df)
colnames(phenotypes_2019.df)

identical_fields <- colnames(phenotypes_2018.df)
sum(phenotypes_2018.df[,identical_fields] != phenotypes_2019.df[,identical_fields])

rm(identical_fields)

```

# Rename matrices

```{r update_variants}

variants.df <- vars.df
genotypes.mx <- gt.mx
phenotypes.df <- phenotypes_2019.df

sum(is.na(phenotypes.df))
sum(phenotypes.df=="")

dim(phenotypes.df)
dim(variants.df)
dim(genotypes.mx)

rm(gt.mx, vars.df, phenotypes_2018.df, phenotypes_2019.df)

```

# Add wecare_phenotypes IDs  

The initial phenotype table includes samples numbers (e.g. 40),  
while the sequencing data include alphanumerical IDs (e.g. 40_S189_L007)  

This chunk adds alphanumerical IDs to the phenotype table as rownames  

```{r phenotypes_rownames}

# Get list of long IDs for all 541 samples
vcf_file="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s04_annotated_vcf/ampliseq_nfe.vcf"
vcfr <- read.vcfR(vcf_file)
gt_739.mx <- extract.gt(vcfr)

dim(gt_739.mx)
gt_739.mx[1:5,1:5]
gt_739.mx[1:5,541:542]
gt_739.mx[1:5,735:739]
wecare_samples.df <- data.frame(long_ids=colnames(gt_739.mx)[1:541])
nfe_samples.df <- data.frame(long_ids=colnames(gt_739.mx)[542:739])

# Split the long IDs for wecare samples
wecare_samples.df <- separate(wecare_samples.df, "long_ids", c("num", "illumina_id", "illumina_lane"), 
                              sep="_", remove=F) 
wecare_samples.df$num <- as.integer(wecare_samples.df$num)
dim(wecare_samples.df)
str(wecare_samples.df)
wecare_samples.df[1:5,]

# Sort wecare_samples.df by sample num (to sync with phenotypes.df)
wecare_samples.df <- wecare_samples.df[order(wecare_samples.df$num),]
wecare_samples.df[1:5,]
phenotypes.df[1:5,1:5]

# Bind add long IDs to phenotypes table
phenotypes.df <- cbind(wecare_samples.df, phenotypes.df)

# Check that numbers are identical
sum(phenotypes.df$num == phenotypes.df$Sample_num)
sum(phenotypes.df$num != phenotypes.df$Sample_num)

# Remove one of the numbers
phenotypes.df <- phenotypes.df[,-2]
dim(phenotypes.df)
colnames(phenotypes.df)
str(phenotypes.df)

# Set rownames
rownames(phenotypes.df) <- phenotypes.df$long_ids

# Check results
genotypes.mx[1:5,1:5]
variants.df[1:5,1:5]
phenotypes.df[1:5,1:5]

dim(genotypes.mx)
dim(variants.df)
dim(phenotypes.df)

# Clean-up
rm(vcf_file, vcfr, gt_739.mx, wecare_samples.df, nfe_samples.df)

```

# Sync phenotypes and genotypes

Remove samples failed sequencing (6 wecare and 1 nfe)  

```{r sync_pheno_geno}

dim(genotypes.mx)
dim(phenotypes.df)

phenotypes.df <- phenotypes.df[colnames(genotypes.mx),]

dim(phenotypes.df)

```

# check data

```{r end_check}

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

table(phenotypes.df$cc)

```

# save_results

```{r save_results}

save.image(paste(base_folder, "s01_add_phenotypes.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
