---
title: "Ampliseq WECARE only: BC8 crude counts - lof and all"
author: "Alexey Larionov"
date: "04 Mar 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

- Select pthogenic variants in BC8, including consensus high-grade missenses
- Calculate crude counts: totals, and separately for cases and controls  
- Fisher test for the crude counts  

Using the old annotations ans simplifyed /inconsistent variant selection  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
#options(width=999)
```

# Start section

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors = F,
        warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

# Files and folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_wo_danish_2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","x01_crude_counts")
scripts_folder <- file.path(base_folder,"scripts","x01_crude_counts")
setwd(scripts_folder)

# Libraries
library(dplyr)
library(VennDiagram)
#library(ggVennDiagram)

```

# Load data

```{r}

load(file.path(data_folder,"s01_crude_counts_per_variant.RData"))

# Update working folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_wo_danish_2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","x01_crude_counts")
scripts_folder <- file.path(base_folder,"scripts","x01_crude_counts")

bc8_genes <- scan(file.path(data_folder,"bc8_genes.txt"),what="character")

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

# Check list of genes
bc8_genes

colnames(variants.df)

```

# Select variants 

## Keep only variants in the selected genes

```{r}

bc8_variants.df <- variants.df[variants.df$SYMBOL %in% bc8_genes,]
dim(bc8_variants.df)

```

## predicted_lof-s as defined by VEP IMPACT

```{r}

predicted_lof <- bc8_variants.df$IMPACT == "HIGH"
sum(predicted_lof)

```

## Select known (likely) pathogenic by ClinVar

```{r}

# Explore ClinVar annotations
bc8_variants.df %>% 
  group_by(CLIN_SIG) %>% 
  summarise(counts=n()) %>% 
  arrange(desc(counts))

# Select known_pathogenic variants
known_pathogenic <- grepl("pathogenic",bc8_variants.df$CLIN_SIG) 
sum(known_pathogenic)

#grepl("risk_factor",bc8_variants.df$CLNSIG)

#bc8_variants.df %>% 
#  group_by(CLNREVSTAT) %>% 
#  summarise(counts=n()) %>% 
#  arrange(desc(counts))

#with_good_support <-
#  variants.df$CLNREVSTAT == "reviewed_by_expert_panel" |
#  variants.df$CLNREVSTAT == "criteria_provided,_multiple_submitters,_no_conflicts" |
#  variants.df$CLNREVSTAT == "criteria_provided,_single_submitter""
#etc

```

## Consensus high grade missenses  

```{r}

# SIFT
bc8_variants.df %>% 
  group_by(SIFT_call) %>% 
  summarise(count=n()) %>% 
  arrange(desc(count))

# PolyPhen
bc8_variants.df %>% 
  group_by(PolyPhen_call) %>% 
  summarise(count=n()) %>% 
  arrange(desc(count))

# Consensus high grade missenses
consensus_high_grade_missenses <- 
  bc8_variants.df$SIFT_call == "deleterious" & !is.na(bc8_variants.df$SIFT_call) & 
  bc8_variants.df$PolyPhen_call == "probably_damaging" & !is.na(bc8_variants.df$PolyPhen_call) 

sum(consensus_high_grade_missenses)

```

## Rare variants

```{r}

# Rare in gnomeD NFE
sum(bc8_variants.df$exac_non_TCGA.AF <= 0.01, na.rm=T)

min(bc8_variants.df$exac_non_TCGA.AF, na.rm=T)
max(bc8_variants.df$exac_non_TCGA.AF, na.rm=T) # some rare variants with MAF<0.01 and AF>0.99

sum(is.na(bc8_variants.df$exac_non_TCGA.AF))

rare_in_exac_non_TCGA <- bc8_variants.df$exac_non_TCGA.AF <= 0.01 | 
  bc8_variants.df$exac_non_TCGA.AF >= 0.99 | 
  is.na(bc8_variants.df$exac_non_TCGA.AF)
sum(rare_in_exac_non_TCGA)

# Rare in dataset
sum(bc8_variants.df$total_af <= 0.01, na.rm=T)
sum(bc8_variants.df$total_af >= 0.99, na.rm=T)

min(bc8_variants.df$total_af, na.rm=T)
max(bc8_variants.df$total_af, na.rm=T) # no rare variants with MAF<0.01 and AF>0.99
sum(is.na(bc8_variants.df$total_af))

rare_in_wecare <- bc8_variants.df$total_af <= 0.01 | bc8_variants.df$total_af >= 0.99
sum(rare_in_wecare)

# Rare in controls
#rare_in_controls <- bc8_variants.df$controls_ac <= 0.05
#sum(rare_in_controls)

futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

sets.ls <- list(exac=bc8_variants.df$SplitVarID[rare_in_exac_non_TCGA],
                wecare=bc8_variants.df$SplitVarID[rare_in_wecare])

z <- venn.diagram(sets.ls,filename=NULL)

grid.draw(z)

#rare <- rare_in_gnomAD_NFE | rare_in_wecare
rare <- rare_in_exac_non_TCGA & rare_in_wecare
#rare <- rare_in_gnomAD_NFE
#rare <- rare_in_controls

sum(rare)

#uncommon <- bc8_variants.df$vep_gnomAD_NFE_AF <= 0.05
#sum(uncommon)

# Clean-up
rm(rare_in_exac_non_TCGA, rare_in_wecare, sets.ls, z)

```

## Known benign or uncertain_significance

```{r}

# Explore ClinVar annotations
bc8_variants.df %>% 
  group_by(CLIN_SIG) %>% 
  summarise(counts=n()) %>% 
  arrange(desc(counts))

# Known benign
known_benign <- grepl("benign",bc8_variants.df$CLIN_SIG)
uncertain_significance <- grepl("uncertain_significance",bc8_variants.df$CLIN_SIG)
sum(known_benign | uncertain_significance)

```

## Combined lof

```{r}

# Functionally important variants - lof
fiv_lof <- (predicted_lof | known_pathogenic) & 
  rare & 
  !(known_benign | uncertain_significance)

sum(fiv_lof)

bc8_fiv_lof.df <- bc8_variants.df[fiv_lof,] %>% 
  select(SplitVarID,SYMBOL,Consequence,IMPACT,CLIN_SIG,PolyPhen_call,SIFT_call,
         exac_non_TCGA.AF,
         total_ac,total_an,total_af,
         controls_ac,controls_an,controls_af,
         cases_ac,cases_an,cases_af)
bc8_fiv_lof.df
#View(bc8_fiv_lof.df)

# Clean-up, known_benign
rm(fiv_lof)

```

## Combined all

```{r}

# Functionally important variants - all 
fiv_all <- (predicted_lof | known_pathogenic | consensus_high_grade_missenses) & 
  rare & 
  !(known_benign | uncertain_significance)

sum(fiv_all)

bc8_fiv_all.df <- bc8_variants.df[fiv_all,] %>% 
  select(SplitVarID,SYMBOL,Consequence,IMPACT,CLIN_SIG,PolyPhen_call,SIFT_call,
         exac_non_TCGA.AF,
         total_ac,total_an,total_af,
         controls_ac,controls_an,controls_af,
         cases_ac,cases_an,cases_af)
bc8_fiv_all.df
#View(bc8_fiv_all.df)

# Clean-up, known_benign
rm(predicted_lof,known_pathogenic,rare,fiv_all,consensus_high_grade_missenses, 
   known_benign)

```

# Aggregated crude counts Fisher test

## LOF

```{r}

# Calculate crude counts
controls_ac <- sum(bc8_fiv_lof.df$controls_ac) 
controls_an <- sum(bc8_fiv_lof.df$controls_an)
controls_af <- controls_ac / controls_an

cases_ac <- sum(bc8_fiv_lof.df$cases_ac)
cases_an <- sum(bc8_fiv_lof.df$cases_an)
cases_af <- cases_ac / cases_an

# Calculate Fisher test
fisher_matrix <- matrix(c(cases_ac, cases_an - cases_ac,
                          controls_ac, controls_an - controls_ac),
                          nrow=2, byrow=T)

colnames(fisher_matrix) <- c("Alt","Ref")
rownames(fisher_matrix) <- c("Cases","Controls")
fisher_matrix 

ft <- fisher.test(fisher_matrix, conf.int=T)

ft

afs <- c(controls_af,cases_af)
names(afs) <- c("Controls","Cases")

cc <- table(phenotypes.df$cc)

mp <- barplot(afs, ylim=c(0,max(afs)+0.0005), 
main=paste("WECARE Ampliseq w/o Danish - LoF\n",cc[1],"controls vs ",cc[2]," cases\nCrude counts Fisher p=",round(ft$p.value,3)), ylab="Mean AF in selected variants")

text(mp,afs+0.0003,round(afs,4))

# Clean-up
rm(controls_ac, controls_an, controls_af,
   cases_ac, cases_an, cases_af,
   afs, mp, fisher_matrix, cc, ft)

```

## All

```{r}

# Calculate crude counts
controls_ac <- sum(bc8_fiv_all.df$controls_ac) 
controls_an <- sum(bc8_fiv_all.df$controls_an)
controls_af <- controls_ac / controls_an

cases_ac <- sum(bc8_fiv_all.df$cases_ac)
cases_an <- sum(bc8_fiv_all.df$cases_an)
cases_af <- cases_ac / cases_an

# Calculate Fisher test
fisher_matrix <- matrix(c(cases_ac, cases_an - cases_ac,
                          controls_ac, controls_an - controls_ac),
                          nrow=2, byrow=T)

colnames(fisher_matrix) <- c("Alt","Ref")
rownames(fisher_matrix) <- c("Cases","Controls")
fisher_matrix 

ft <- fisher.test(fisher_matrix, conf.int=T)

ft

afs <- c(controls_af,cases_af)
names(afs) <- c("Controls","Cases")

cc <- table(phenotypes.df$cc)

mp <- barplot(afs, ylim=c(0,max(afs)+0.0005), 
main=paste("WECARE Ampliseq w/o Danish - All\n",cc[1],"controls vs ",cc[2]," cases\nCrude counts Fisher p=",round(ft$p.value,3)), ylab="Mean AF in selected variants")

text(mp,afs+0.0003,round(afs,4))

# Clean-up
rm(controls_ac, controls_an, controls_af,
   cases_ac, cases_an, cases_af,
   afs, mp, fisher_matrix, cc, ft)

```

# Carriers of the selected variants

```{r}

genotypes_bc8.mx <- genotypes.mx[bc8_fiv_all.df$SplitVarID,]
dim(genotypes_bc8.mx)

carriers.udf <- function(x){paste(names(x)[which(x>0)], collapse=",")}
x <- apply(genotypes_bc8.mx,1,carriers.udf)
x

rm(x, carriers.udf)

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(bc8_variants.df)
dim(bc8_fiv_lof.df)
dim(bc8_fiv_all.df)

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

```

# Save results

```{r}

save.image(file.path(data_folder,"s02_crude_counts_bc8_lof_all.RData"))

```

# Final section

```{r}

ls()
sessionInfo()
Sys.time()
gc()

```
